### ISSUE TITLE
Test CSS selectors against HTML data

### ISSUE THREAD

**Issue by dag** [dag/attest#43]
Parse with lxml.html, match with .cssselect().

API?

```
Assert(html).css(selectors)
```

**Comment by DasIch** [comment]
.css() should return an iterable over the elements(wrapped in Assert) the selector matched. I would leave everything else to the user.

**Comment by dag** [comment]
I was thinking it only tests that the selectors match. Do you think it could be useful to run tests on the matches themselves?

I tend to think that testing for _content_ in HTML isn't very useful. I suppose it's still useful to test things like how _many_ matches there are for a selector… "If I add three Xs to the DB and GET this URL, I expect three 'article.x'"

**Comment by DasIch** [comment]
Running tests on the matches themselves can certainly be useful, even going so far as checking the content.

I think a very important part of testing is that you attempt to remain ignorant of the implementation and architecture, the less you know about it the better.

If you built a car every part it is built out of is thoroughly and redundantly tested, that doesn't mean you don't have to test the car.

So if you test your webapp you want to make sure that if you post certain content that content appears in the resulting html you see as you want to see it, even though you already tested all those Models, datastructures, helper functions and whatnot.

**Comment by dag** [comment]
Agreed: testing started to make sense to me when I read wycats' promoting testing the _public_ interfaces.

Anyway because of Assert's `__nonzero__` it should be simple to do the test I originally intended anyway:

```
response = Assert(client.get(...))
assert response.data.css('#id')
```

**Comment by dag** [comment]
Should `Assert.json` be an instance method/property to be more compatible with this API?

**Comment by DasIch** [comment]
I think so, however I think Assert.isinstance/issubclass should be instance methods as well or at least a variant e.g. Assert.isinstanceof.

**Comment by dag** [comment]
My reasoning with making _those_ staticmethods was to make them "feel" more like normal Python:

```
isinstance(x, y)
Assert.isinstance(x, y)
```

Alien:

```
Assert(x).isinstance(y)
```

A benefit to making those instance methods would be that you could use them with attribute proxying, but I'm starting to realize that said proxying is probably a bad thing to begin with, and I'll likely be adding warnings against it in the docs.

**Comment by dag** [comment]
Another potential API would be

```
Assert(data, json=True) == dict(...)
```

Shrug.

**Comment by DasIch** [comment]
`Assert(..., json=True)` is not a good idea, do you want to support `yaml`, `xml` and other container formats like this as well? The data processing part has nothing to do with the `Assert` object itself and therefore shouldn't be part of it's constructor.

**Comment by dag** [comment]
True, I didn't like that one myself either.

**Comment by dag** [comment]
Assert.css, closed by 87449fb84d5806879ca8cdb9aa00f13a5de6138d
